Bottom: 3e90b33543a7ad7f7e207d2b8560406cce1239f4
Top:    cb13e87a22ec7057438542d7607ee64dc5e68834
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-04-23 16:56:13 +0100

draw.c: Change DrawTop() to DrawPPV(), and teach it to work for any layer

(PPV stands for Pins, Pads and Vias). Extend its functionality to work
for pins and vias on any layer.


---

diff --git a/src/draw.c b/src/draw.c
index 7380beb..2a097d9 100644
--- a/src/draw.c
+++ b/src/draw.c
@@ -83,7 +83,7 @@ static const BoxType *clip_box = NULL;
  */
 static void Redraw (bool, BoxTypePtr);
 static void DrawEverything (BoxTypePtr);
-static void DrawTop (const BoxType *);
+static void DrawPPV (int group, const BoxType *);
 static int DrawLayerGroup (int, const BoxType *);
 static void DrawPinOrViaLowLevel (PinTypePtr, bool);
 static void DrawPlainPin (PinTypePtr, bool);
@@ -312,14 +312,15 @@ static void
 PrintAssembly (const BoxType * drawn_area, int side_group, int swap_ident)
 {
   int save_swap = SWAP_IDENT;
+  int side = swap_ident ? SOLDER_LAYER : COMPONENT_LAYER;
 
   gui->set_draw_faded (Output.fgGC, 1);
-  SWAP_IDENT = swap_ident;
   DrawLayerGroup (side_group, drawn_area);
-  DrawTop (drawn_area);
+  DrawPPV (side_group, drawn_area);
   gui->set_draw_faded (Output.fgGC, 0);
 
   /* draw package */
+  SWAP_IDENT = swap_ident;
   DrawSilk (swap_ident,
 	    swap_ident ? solder_silk_layer : component_silk_layer,
 	    drawn_area);
@@ -404,7 +405,7 @@ DrawEverything (BoxTypePtr drawn_area)
 
   /* Draw pins, pads, vias below silk */
   if (gui->gui)
-    DrawTop (drawn_area);
+    DrawPPV (SWAP_IDENT ? solder : component, drawn_area);
   else
     {
       HoleCountStruct hcs;
@@ -569,17 +570,30 @@ pad_callback (const BoxType * b, void *cl)
  * draws pins pads and vias
  */
 static void
-DrawTop (const BoxType * screen)
+DrawPPV (int group, const BoxType * screen)
 {
-  int side = SWAP_IDENT ? SOLDER_LAYER : COMPONENT_LAYER;
+  int component_group = GetLayerGroupNumberByNumber (component_silk_layer);
+  int solder_group = GetLayerGroupNumberByNumber (solder_silk_layer);
 
   if (PCB->PinOn || doing_assy)
     {
       /* draw element pins */
       r_search (PCB->Data->pin_tree, screen, NULL, pin_callback, NULL);
+
       /* draw element pads */
-      r_search (PCB->Data->pad_tree, screen, NULL, pad_callback, &side);
+      if (group == component_group)
+        {
+          side = COMPONENT_LAYER;
+          r_search (PCB->Data->pad_tree, drawn_area, NULL, pad_callback, &side);
+        }
+
+      if (group == solder_group)
+        {
+          side = COMPONENT_LAYER;
+          r_search (PCB->Data->pad_tree, drawn_area, NULL, pad_callback, &side);
+        }
     }
+
   /* draw vias */
   if (PCB->ViaOn || doing_assy)
     {
